### PR TITLE
Graph Bert training

### PR DESCRIPTION
**Pull request by L1aoXingyu**
本 PR 解决下面的问题

- [x] AdamW with weigth_decay_exclude
- [x] PolynomialDecayed lr scheduler
- [x] warmup lr
- [x] clip gradients
- [x] consistent DDP training
- [x] AMP training
- [ ] 完整 Pre-training 实验 
- [x] SQuAD 验证

### CODE REVIEW

**Comment by L1aoXingyu** [comment]
完成 adamW 支持，和 lazy 的 loss 曲线如下
lazy_graph_adamw_loss

**Comment by CPFLAME** [comment]
### bert nn grpah与lazy对齐  复现更新

完成polynomial_scheduler和warmuplr的添加，从sgd的对齐结果上看没有问题

目前在sgd上可以完美对齐，但是在adam和adamw上有细微偏差。

### sgd 实验

- 优化器：sgd
- batch size：32
- lr：0.02
lazy_graph_sgd

### adamw实验
注意和星宇的实验配置不一样的地方为本实验lr=1e-3而星宇的实验lr=1e-4，所以在loss曲线上会有差别
- 优化器：adamw
- batch size：32
- lr：1e-3
lazy_graph_adamw

### Adam实验
- 优化器：adam
- batch size：32
- lr：1e-3
lazy_graph_adam

**Comment by CPFLAME** [comment]
### 单卡上添加consistent和sbp
对齐通过，
注意需要将oneflow源码中[normalization.py](https://github.com/Oneflow-Inc/oneflow/blob/9764496c089e48d928eb72494a26318545fff4c7/python/oneflow/nn/modules/normalization.py#L262)代码262行`if x.device == flow.device("cpu"):` 替换成`if not x.is_cuda:`

- 优化器：sgd
- batch size：32
- lr：0.02
lazy_graph_sgd_amp_consistent

**Comment by CPFLAME** [comment]
### loss对齐曲线 graph && lazy
oneflow版本：`0.5.0.dev20210923+cu111`
显卡型号：`GeForce RTX 2080 Ti`

基本配置
- `lr=1e-4`, 
- `batch_size_per_gpu=32`
- `单机4卡运行` 

所有trick
- adamw optimizer with weigth_decay_exclude
- warm up + polynominal lr scheduler 
- amp
- consistent + sbp
- clip gradients


image

**Comment by Ldpe2G** [comment]
README.md 也需要更新，每个 .sh 脚本都可以简单介绍一下是做什么的

还有命令行参数的意思也可以简单列出来，可以参考 resnet50 https://github.com/Oneflow-Inc/models/tree/main/Vision/classification/image/resnet50#explaination-of-command-parameters-for-trainpy-script

File: NLP/BERT/modeling.py [hunk 1]
```diff
@@ -302,9 +310,21 @@ def get_extended_attention_mask(
         output = flow.cast(attention_mask, dtype=flow.float32)
         output = flow.reshape(output, [-1, 1, to_seq_length])
         # broadcast `from_tensor` from 2D to 3D
-        zeros = flow.zeros(
-            (from_seq_length, to_seq_length), dtype=flow.float32, device=output.device
-        )
+
```
Comment (Ldpe2G) [inline NLP/BERT/modeling.py:313]: 这里改成 output.expand 就行了？

File: NLP/BERT/modeling.py [hunk 2]
```diff
@@ -233,7 +236,12 @@ def __init__(self, hidden_size):
     def forward(self, hidden_states):
         """Just "pool" the model by simply taking the [CLS] token corresponding to the first token.
         """
+        hidden_size = hidden_states.shape[-1]
+        # NOTE: Don't use [:, 0] when it's consistent tensor, use flow.slice, otherwise its sbp will be
+        # converted to `broadcast`
+        # first_token_tensor = flow.slice(hidden_states, [[None, None, None], [0, 1, 1]])
```
Comment (Ldpe2G) [inline NLP/BERT/modeling.py:242]: 现在最新的master分支也还是有这个问题吗，如果有的话，需要加个  is_consistent 的判断

File: NLP/BERT/run_eager_pretrain.sh [hunk 3]
```diff
@@ -0,0 +1,28 @@
...
+_DEVICE_NUM_PER_NODE=1
+_MASTER_ADDR=127.0.0.1
+_NUM_NODES=1
+_NODE_RANK=0
+
+
+_LR=0.02
+_BATCH_SIZE_PER_GPU=32
+# train_data_dir=./wiki_ofrecord_seq_len_128_1part_repeat4
+# train_data_dir=/dataset/bert/of_wiki_seq_len_128
```
Comment (Ldpe2G) [inline NLP/BERT/run_eager_pretrain.sh:12]: 这些注释可以删掉了

File: NLP/BERT/run_eager_pretrain.sh [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+train_data_dir=./wiki_ofrecord_seq_len_128_4part
+LOGFILE=./bert_graph_pretrain.log
+
+# export ONEFLOW_DEBUG_MODE=1
```
Comment (Ldpe2G) [inline NLP/BERT/run_eager_pretrain.sh:16]: 这个也删了

File: NLP/BERT/run_eager_pretrain.sh [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+export PYTHONUNBUFFERED=1
+# export CUDA_VISIBLE_DEVICES=1
```
Comment (Ldpe2G) [inline NLP/BERT/run_eager_pretrain.sh:18]: 这个也删了

File: NLP/BERT/modeling.py (hunk 2 above)
Comment (L1aoXingyu) [reply NLP/BERT/modeling.py:242]: 最新的分支可以用 `[:, 0]`, slice 那个应该也修了

File: NLP/BERT/run_eager_pretrain.sh [hunk 6]
```diff
@@ -0,0 +1,28 @@
+#!/bin/bash
+
+_DEVICE_NUM_PER_NODE=1
```
Comment (Ldpe2G) [inline NLP/BERT/run_eager_pretrain.sh:3]: ddp 有测过多卡吗

Comment (CPFLAME) [reply NLP/BERT/run_eager_pretrain.sh:3]: ddp的loss对齐测的就是多卡。https://github.com/Oneflow-Inc/models/issues/187#issuecomment-920798546

File: NLP/BERT/run_eager_pretraining.py [hunk 7]
```diff
@@ -0,0 +1,322 @@
...
+            desc="bert pretrain",
+            print_steps=args.loss_print_every_n_iters,
+            batch_size=args.train_batch_size,
+            keys=["total_loss", "mlm_loss", "nsp_loss", "pred_acc"],
+        )
+
+        # Train
+        bert_model.train()
+
+        for step in range(300):
```
Comment (Ldpe2G) [inline NLP/BERT/run_eager_pretraining.py:293]: 这里应该是从 `train_data_loader` 获取一个epoch的迭代次数
可以参考 resnet50 ，新增了一个 `total_batch_size` 的参数
https://github.com/Oneflow-Inc/models/blob/main/Vision/classification/image/resnet50/utils/ofrecord_data_utils.py#L15

File: NLP/BERT/run_pretraining.py [hunk 8]
```diff
@@ -0,0 +1,549 @@
...
+    def str2bool(v):
+        if v.lower() in ("yes", "true", "t", "y", "1"):
+            return True
+        elif v.lower() in ("no", "false", "f", "n", "0"):
+            return False
+        else:
+            raise argparse.ArgumentTypeError("Unsupported value encountered.")
+
+    parser = argparse.ArgumentParser()
+
```
Comment (Ldpe2G) [inline NLP/BERT/run_pretraining.py:141]: 是不是可以把 `run_eager_pretraining.py` 和 `run_pretraining.py` 参数定义都放到一个文件里，应该有不少是重复的

File: NLP/BERT/modeling.py (hunk 2 above)
Comment (Ldpe2G) [reply NLP/BERT/modeling.py:242]: 好的

File: NLP/BERT/run_eager_pretrain.sh (hunk 6 above)
Comment (Ldpe2G) [reply NLP/BERT/run_eager_pretrain.sh:3]: 好的

File: NLP/BERT/modeling.py (hunk 1 above)
Comment (L1aoXingyu) [reply NLP/BERT/modeling.py]: 已经修改了

File: NLP/BERT/README.md [hunk 9]
```diff
@@ -0,0 +1,260 @@
...
+```
+
+After fine-tuning, you can get the evaluation results as follow
+```
+{"exact_match": 82.57332071901608, "f1": 89.63355726606137}
+```
+
+### 4. Inference 
+
+You just provide a simple inference script in `example/infer.sh`, you can also run inference with following command
```
Comment (Ldpe2G) [inline NLP/BERT/README.md:245]: 应该是  `We just  xx` ?

File: NLP/BERT/README.md [hunk 10]
```diff
@@ -0,0 +1,260 @@
...
+
+```
+MODEL_PATH="checkpoints/epoch_9_val_acc_0.554688"
```
Comment (Ldpe2G) [inline NLP/BERT/README.md:248]: 这个模型怎么得到的

File: NLP/BERT/README.md [hunk 11]
```diff
@@ -0,0 +1,260 @@
...
+  --model_path $MODEL_PATH
+```
+
+## Utils
+
+### 1. Lazy Mode Loading
+Bert lazy model can be loaded in `nn.Module` using the function `load_params_from_lazy` from `compare_lazy_outputs.py`.
+
+### 2. Draw Loss Curve
+You can make use of `draw_loss_curve.py` to draw loss curve for aligning different train mode.
```
Comment (Ldpe2G) [inline NLP/BERT/README.md:260]: `compare_lazy_outputs.py` 和 `draw_loss_curve.py` 这两个脚本应该要放到 utils 目录下

File: NLP/BERT/README.md (hunk 10 above)
Comment (L1aoXingyu) [reply NLP/BERT/README.md]: 我留了一个 lazy 转 TensorFlow 模型的链接

**Review by ShawnXuan — COMMENTED** [review]
1. argparse 出现在好几个文件中，比如config.py train_config.py run_squad.py。建议只用一个config.py文件，包括了所有公用的参数。预训练和下游任务中如果需要自己的参数，可以在run_xxx.py中添加自己特有的参数。
2.  BERT的预训练还是以steps作为主循环，不建议用epoch，用steps计算总的训练token数比较容易，epoch还受到数据集大小的影响。目前ofrecord沿用的是tfrecord，数据集中sample的数量是固定的。未来可能会使用只做了tokenized的数据集，预训练训练任务或者下游任务的数据根据需要按batch生成，这样的话一个epoch的大小更加不好定义了。
3. 建议增加一个100轮且每一轮打印一次loss的对比记录。

**Comment by L1aoXingyu** [comment]
> 1. argparse 出现在好几个文件中，比如config.py train_config.py run_squad.py。建议只用一个config.py文件，包括了所有公用的参数。预训练和下游任务中如果需要自己的参数，可以在run_xxx.py中添加自己特有的参数。
> 2. BERT的预训练还是以steps作为主循环，不建议用epoch，用steps计算总的训练token数比较容易，epoch还受到数据集大小的影响。目前ofrecord沿用的是tfrecord，数据集中sample的数量是固定的。未来可能会使用只做了tokenized的数据集，预训练训练任务或者下游任务的数据根据需要按batch生成，这样的话一个epoch的大小更加不好定义了。
> 3. 建议增加一个100轮且每一轮打印一次loss的对比记录。

关于第二点按照 step 作为主循环，对于同一个数据集，如果用户需要修改 batch size，那么也需要同时修改 steps，这样并不是很方便，如果按照 epoch 的概念来说，修改了 batch size 之后，step 是通过 epoch * (dataset_size) / batch_size 去自动计算的，就不需要用户手动去修改了；如果换了一个数据集，在数据集大小是未知的情况下直接设定 steps 也比较困难，用户可能就是想遍历整个数据集 3 次，这样直接设定 epoch=3 就可以完成这个需求，如果设定 step，用户还需要自己去统计一些信息。

**Comment by ShawnXuan** [comment]
> 关于第二点按照 step 作为主循环，对于同一个数据集，如果用户需要修改 batch size，那么也需要同时修改 steps，这样并不是很方便，如果按照 epoch 的概念来说，修改了 batch size 之后，step 是通过 epoch * (dataset_size) / batch_size 去自动计算的，就不需要用户手动去修改了；如果换了一个数据集，在数据集大小是未知的情况下直接设定 steps 也比较困难，用户可能就是想遍历整个数据集 3 次，这样直接设定 epoch=3 就可以完成这个需求，如果设定 step，用户还需要自己去统计一些信息。

NLP一般是以训练了多少个token数(train samples)为目标，再退一步是train iters，再退一步才是train epochs。数据集是原始的文本或者经过tokenized的id，任务不同，即使同一个数据集能够产生的sample数量也是不同的，而且事先未毕知晓，甚至不同的epoch之间的数量也不同，所以我不建议用epoch。

另外原始BERT和GPT里也都没有用epoch。

**Comment by L1aoXingyu** [comment]
> > 关于第二点按照 step 作为主循环，对于同一个数据集，如果用户需要修改 batch size，那么也需要同时修改 steps，这样并不是很方便，如果按照 epoch 的概念来说，修改了 batch size 之后，step 是通过 epoch * (dataset_size) / batch_size 去自动计算的，就不需要用户手动去修改了；如果换了一个数据集，在数据集大小是未知的情况下直接设定 steps 也比较困难，用户可能就是想遍历整个数据集 3 次，这样直接设定 epoch=3 就可以完成这个需求，如果设定 step，用户还需要自己去统计一些信息。
> 
> NLP一般是以训练了多少个token数(train samples)为目标，再退一步是train iters，再退一步才是train epochs。数据集是原始的文本或者经过tokenized的id，任务不同，即使同一个数据集能够产生的sample数量也是不同的，而且事先未毕知晓，甚至不同的epoch之间的数量也不同，所以我不建议用epoch。
> 
> 另外原始BERT和GPT里也都没有用epoch。

好的，xuan 哥，我们再看看 hugging face 里面是如何处理的，如果都是按 step 为单位做训练，就改为 step

**Comment by L1aoXingyu** [comment]
看了一下 huggingface 的代码，支持两种方式传参
https://github.com/huggingface/transformers/blob/a21ee1f99003f49f411b3b62a087c4da5d839843/examples/pytorch/language-modeling/run_clm_no_trainer.py#L132-L138

https://github.com/huggingface/transformers/blob/a21ee1f99003f49f411b3b62a087c4da5d839843/examples/pytorch/language-modeling/run_clm_no_trainer.py#L440-L444

要不我们也按照这个方式传参？@ShawnXuan

**Comment by CPFLAME** [comment]
### 100轮loss的对比记录

image

**Review by Ldpe2G — APPROVED** [review]

File: NLP/BERT/run_pretraining.py [hunk 12]
```diff
@@ -0,0 +1,490 @@
...
+    assert id(bert_model.cls.predictions.decoder.weight) == id(
+        bert_model.bert.embeddings.word_embeddings.weight
+    )
+
+    ns_criterion = nn.CrossEntropyLoss(reduction="mean")
+    mlm_criterion = nn.CrossEntropyLoss(reduction="none")
+
+    if args.use_consistent:
+        placement = flow.placement("cuda", {0: range(flow.env.get_world_size())})
```
Comment (chengtbf) [inline NLP/BERT/run_pretraining.py:294]: 这里的 placement 这么配置是错误的。仅在 单机 下有效。多机（比如 2机8卡），这里的 placement会变成：

{0:[0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15]}

但实际上需要的是：

{0:[0,1,2,3,4,5,6,7],  1:[0,1,2,3,4,5,6,7]}

有一个简写的方式（如果希望  placement 是所有的 rank）

```python
flow.env.all_device_placement("cuda")
```

File: NLP/BERT/utils/ofrecord_data_utils.py [hunk 13]
```diff
@@ -11,16 +11,33 @@ def __init__(
...
+        self.use_consistent = consistent
+        self.data_part_num = data_part_num
+
+        if self.use_consistent:
+            self.world_size = flow.env.get_world_size()
+            if data_part_num < self.world_size:
+                self.placement = flow.placement("cpu", {0: [0]})
+                self.sbp = flow.sbp.broadcast
+            else:
+                self.placement = flow.placement("cpu", {0: range(self.world_size)})
```
Comment (chengtbf) [inline NLP/BERT/utils/ofrecord_data_utils.py:30]: 这里的配置也是错误的。相关的 placement 都需要检查是否只写了这种单卡情形。

应该改成： 

flow.env.all_device_placement("cpu")

File: NLP/BERT/utils/ofrecord_data_utils.py [hunk 14]
```diff
@@ -53,6 +70,23 @@ def forward(self):
         masked_lm_positions = self.blob_confs["masked_lm_positions"](data_record)
         masked_lm_weights = self.blob_confs["masked_lm_weights"](data_record)
 
+        if self.use_consistent and self.data_part_num < self.world_size:
+            placement = flow.placement("cpu", {0: range(self.world_size)})
```
Comment (chengtbf) [inline NLP/BERT/utils/ofrecord_data_utils.py:74]: flow.env.all_device_placement("cpu")
